README.md
# mrcnpdlk/url-search-parser
[![Latest Stable Version](https://img.shields.io/github/release/mrcnpdlk/url-search-parser.svg)](https://packagist.org/packages/mrcnpdlk/url-search-parser)
[![Latest Unstable Version](https://poser.pugx.org/mrcnpdlk/url-search-parser/v/unstable.png)](https://packagist.org/packages/mrcnpdlk/url-search-parser)
[![Total Downloads](https://img.shields.io/packagist/dt/mrcnpdlk/url-search-parser.svg)](https://packagist.org/packages/mrcnpdlk/url-search-parser)
[![Monthly Downloads](https://img.shields.io/packagist/dm/mrcnpdlk/url-search-parser.svg)](https://packagist.org/packages/mrcnpdlk/url-search-parser)
[![License](https://img.shields.io/packagist/l/mrcnpdlk/url-search-parser.svg)](https://packagist.org/packages/mrcnpdlk/url-search-parser)
[![FOSSA Status](https://app.fossa.io/api/projects/git%2Bgithub.com%2Fmrcnpdlk%2Furl-search-parser.svg?type=shield)](https://app.fossa.io/projects/git%2Bgithub.com%2Fmrcnpdlk%2Furl-search-parser?ref=badge_shield)

------

[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/mrcnpdlk/url-search-parser/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/mrcnpdlk/url-search-parser/?branch=master)
[![Code Coverage](https://scrutinizer-ci.com/g/mrcnpdlk/url-search-parser/badges/coverage.png?b=master)](https://scrutinizer-ci.com/g/mrcnpdlk/url-search-parser/?branch=master)
[![Build Status](https://scrutinizer-ci.com/g/mrcnpdlk/url-search-parser/badges/build.png?b=master)](https://scrutinizer-ci.com/g/mrcnpdlk/url-search-parser/build-status/master)

[![Code Climate](https://codeclimate.com/github/mrcnpdlk/url-search-parser/badges/gpa.svg)](https://codeclimate.com/github/mrcnpdlk/url-search-parser) 
[![Issue Count](https://codeclimate.com/github/mrcnpdlk/url-search-parser/badges/issue_count.svg)](https://codeclimate.com/github/mrcnpdlk/url-search-parser)


[![Build Status](https://travis-ci.com/mrcnpdlk/url-search-parser.svg?branch=master)](https://travis-ci.com/mrcnpdlk/url-search-parser)

# Contents

This boundle has been created for parsing advanced queries to easy-to-use objects. 

Based on https://www.vinaysahni.com/best-practices-for-a-pragmatic-restful-api#advanced-queries

1. [Installation](#installation)
2. [Supported parameters](#supported-parameters)
   1. [Sort](#sort)
   2. [Filter](#filter)
   3. [Limit](#limit)
   4. [Page](#page)
   5. [Phrase](#phrase)
   6. [Others](#other-params)
3. [Usage](#usage)


## Installation

Install the latest version with [composer](https://packagist.org/packages/mrcnpdlk/teryt-api)
```bash
composer require mrcnpdlk/url-search-parser
```
## Supported parameters

```php
$oParser = new \Mrcnpdlk\Lib\UrlQueryParser\RequestParser($query); 
```

### Sort

Generic parameter `sort` can be used to describe sorting rules. Accommodate complex sorting requirements by letting the sort parameter take in list of comma separated fields, each with a possible unary negative to imply descending sort order. Let's look at some examples: 

- `GET /messages?sort=-createDate` - Retrieves a list of messages in descending order of createDate;
- `GET /messages?sort=-type,createDate` - Retrieves a list of messages in descending order of type. Within a specific type, older messages are ordered first;

```php
/**
 * @var $oSort \Mrcnpdlk\Lib\UrlQueryParser\Criteria\Sort
 */
$oSort = $oParser->getSort();
```

### Filter

Filtering is more complex then sorting. Array notation is used. Let's look at some examples: 

- `GET /messages?filter[created][lt]=2018-06-01` - Retrieves a list of messages where createDate is lower than 2018-06-01;
- `GET /messages?filter[type][in]=urgent,warning,error` - Retrieves a list of messages where type is urgent, warning or error;
- `GET /messages?filter[type][null]` - Retrieves a list of messages where type is NULL;

Allowed operators: `eq`,`lt`,`lte`,`gt`,`gte`,`like`,`llike`,`rlike`,`in`,`notin`,`null`,`notnull`,`regexp`

In case using not allowed operator `mrcnpdlk\Lib\UrlQueryParser\Exception\InvalidParamException` is thrown.

```php
/**
 * @var $oFilter \Mrcnpdlk\Lib\UrlQueryParser\Criteria\Filter
 */
$oFilter = $oParser->getFilter();
```

### Limit

Example:

- `GET /messages?limit=20` - limitation;

```php
/**
 * @var $iLimit integer|null
 */
$iLimit = $oParser->getLimit(10); // in NULL default=10
```

### Page

Example:

- `GET /messages?page=1` - pagination. Should be used with `limit` parameter;

```php
/**
 * @var $iPage integer|null
 */
$iPage = $oParser->getPage(1); // in NULL default=1
```

### Phrase

Example:

- `GET /messages?phrase=foo` - for easier filtering;

```php
/**
 * @var $sPhrase string|null
 */
$sPhrase = $oParser->getPhrase(); // if not set NULL ir returned
```

### Other params

Use `getQueryParam()` method.

Example:

- `GET /messages?foo=bar&baz=5` - additional params;

```php
$sFoo = $oParser->getQueryParam('foo','string'); // return 'bar'
$sBaz = $oParser->getQueryParam('baz','int'); // return 5
```



## Usage

```php
// Two ways to get `query` argument for RequestParser constructor:
$query = parse_url($url, PHP_URL_QUERY); // OR
$query = $_SERVER['QUERY_STRING'];

$oParser = new \Mrcnpdlk\Lib\UrlQueryParser\RequestParser($query); 
```



### Example

```php
$url = 'https://api.expample.com?sort=id,-name&filter[isFoo][eq]=1&filter[age][gt]=12&page=3&limit=10&offset=20';
$query =  parse_url($url, PHP_URL_QUERY);
$oParser = new \Mrcnpdlk\Lib\UrlQueryParser\RequestParser($query);

print_r($oParser->getSort()->toArray());
print_r($oParser->getFilter()->toArray());
print_r($oParser->getLimit());
print_r($oParser->getPage());
print_r($oParser->getPhrase());
```

Result

```cassandra
Array
(
    [0] => mrcnpdlk\Lib\UrlQueryParser\Criteria\SortParam Object
        (
            [param] => id
            [direction] => ASC
        )

    [1] => mrcnpdlk\Lib\UrlQueryParser\Criteria\SortParam Object
        (
            [param] => name
            [direction] => DESC
        )

)
Array
(
    [0] => mrcnpdlk\Lib\UrlQueryParser\Criteria\FilterParam Object
        (
            [param] => isFoo
            [operator] => eq
            [sqlOperator] => =
            [value] => 1
        )

    [1] => mrcnpdlk\Lib\UrlQueryParser\Criteria\FilterParam Object
        (
            [param] => age
            [operator] => gt
            [sqlOperator] => >
            [value] => 12
        )

)
10 // limit
3  // page
20 // offset
```



## Running the tests

```bash
./vendor/bin/phpunit
```

## Authors

* **Marcin Pudełek** - *Initial work* - [mrcnpdlk](https://github.com/mrcnpdlk)

See also the list of [contributors](https://github.com/mrcnpdlk/url-search-parser/graphs/contributors) who participated in this project.

## License

Copyright (c) 2018 Marcin Pudełek / mrcnpdlk

This project is licensed under the MIT License - see the [LICENSE](https://github.com/mrcnpdlk/url-search-parser/blob/master/LICENSE) file for details


[![FOSSA Status](https://app.fossa.io/api/projects/git%2Bgithub.com%2Fmrcnpdlk%2Furl-search-parser.svg?type=large)](https://app.fossa.io/projects/git%2Bgithub.com%2Fmrcnpdlk%2Furl-search-parser?ref=badge_large)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [v2.7.0] - 2022-05-23
### Added:
- `removeByParamName` method in Sort object
### Changed:
- PhpCsFixer update
## [v2.6.0] - 2020-09-30
### Added:
- `removeByParam` method
## [v2.5.1] - 2020-09-30
### Changed:
- `isRegexp` -> `isWhereRegexp`
- doc
## [v2.5.0] - 2020-09-30
### Added:
- `rlike`, `llike`, `regexp` filters
## [v2.4.0] - 2020-03-19
### Changed:
- composer up
### Added:
- rawurldecode()
## [v2.3.0] - 2019-11-25
### Changed
- empty phrase means NULL
## [v2.2.0] - 2019-11-03
### Added:
- method `getQuery` - issue [[Feature Request] support http_build_query](https://github.com/mrcnpdlk/url-search-parser/issues/8)
- method `setQueryParam`
### Fixed
- removing specified query param will throw an InvalidParamException

## [v2.1.0] - 2019-09-06
### Added
- FOSSA check
- phpstan warnings fix
- throw `DuplicateParamException` for duplicate sort param
- test, PHPUnit 8.0
## [v2.0.0] - 2019-03-29
### Changed
- namespace change `mrcnpdlk`->`Mrcnpdlk`
### Added
- operator: `not` for filters
## [v1.5.0] - 2019-01-29
### Added:
- setters
## [v1.4.0] - 2019-01-18
### Added
- new method: `replaceParam()`
- tests
## [v1.3.0] - 2019-01-17
### Added
- new method: `getByParam()`
### Changed
- tests
## [v1.2.0] - 2018-10-05
### Added
- new method: `getQuery()`, `getQueryHash()`
- README
## [v1.1.3] - 2018-09-31
### Added
- null & notnull operator support
- BOOLEAN support 
## [v1.1.2] - 2018-07-20
### Added
- new method: `removeQueryParam()`
## [v1.1.1] - 2018-06-17
### Changed
- `getQueryParam()`: type can be NULL, convert AS IS
## [v1.1.0] - 2018-06-17
### Added
- new method: `getQueryParam()`
## [v1.0.0] - 2018-06-17
### Added
- new params support: `like`
- CI integration
- namespace change
- tests
## [v0.2.0] - 2018-06-16
### Added
- new params support: `phrase`,`page`,`offset`
## [v0.1.2] - 2018-06-16
### Fixed
- filter params can be empty array
## [v0.1.1] - 2018-06-16
### Fixed
- `print_r`

## [0.1.0] - 2018-06-16

[Unreleased]: https://github.com/mrcnpdlk/url-search-parser/compare/v2.3.0...devel
[v2.3.0]: https://github.com/mrcnpdlk/url-search-parser/compare/v2.2.0...v2.3.0
[v2.2.0]: https://github.com/mrcnpdlk/url-search-parser/compare/v2.1.0...v2.2.0
[v2.1.0]: https://github.com/mrcnpdlk/url-search-parser/compare/v2.0.0...v2.1.0
[v2.0.0]: https://github.com/mrcnpdlk/url-search-parser/compare/v1.5.0...v2.0.0
[v1.5.0]: https://github.com/mrcnpdlk/url-search-parser/compare/v1.4.0...v1.5.0
[v1.4.0]: https://github.com/mrcnpdlk/url-search-parser/compare/v1.3.0...v1.4.0
[v1.3.0]: https://github.com/mrcnpdlk/url-search-parser/compare/v1.2.0...v1.3.0
[v1.2.0]: https://github.com/mrcnpdlk/url-search-parser/compare/v1.1.3...v1.2.0
[v1.1.3]: https://github.com/mrcnpdlk/url-search-parser/compare/v1.1.2...v1.1.3
[v1.1.2]: https://github.com/mrcnpdlk/url-search-parser/compare/v1.1.1...v1.1.2
[v1.1.1]: https://github.com/mrcnpdlk/url-search-parser/compare/v1.1.0...v1.1.1
[v1.1.0]: https://github.com/mrcnpdlk/url-search-parser/compare/v1.0.0...v1.1.0
[v1.0.0]: https://github.com/mrcnpdlk/url-search-parser/compare/v0.2.0...v1.0.0
[v0.2.0]: https://github.com/mrcnpdlk/url-search-parser/compare/v0.1.2...v0.2.0
[v0.1.2]: https://github.com/mrcnpdlk/url-search-parser/compare/v0.1.1...v0.1.2
[v0.1.1]: https://github.com/mrcnpdlk/url-search-parser/compare/v0.1.0...v0.1.1
[v0.1.0]: https://github.com/mrcnpdlk/url-search-parser/releases/tag/v0.1.0


src/Criteria/Filter.php
<?php
/**
 * Created by Marcin.
 * Date: 16.06.2018
 * Time: 19:05
 */
declare(strict_types=1);

namespace Mrcnpdlk\Lib\UrlSearchParser\Criteria;

use function array_key_exists;
use ArrayIterator;
use function gettype;
use function in_array;
use function is_array;
use function is_string;
use IteratorAggregate;
use Mrcnpdlk\Lib\UrlSearchParser\Exception\InvalidParamException;
use Traversable;

/**
 * Class Filter
 */
class Filter implements IteratorAggregate
{
    public const DELIMITER = ',';

    public const PARAM_EQ         = 'eq';
    public const PARAM_GT         = 'gt';
    public const PARAM_GTE        = 'gte';
    public const PARAM_LT         = 'lt';
    public const PARAM_LTE        = 'lte';
    public const PARAM_LIKE       = 'like';
    public const PARAM_LIKE_RIGHT = 'rlike';
    public const PARAM_LIKE_LEFT  = 'llike';
    public const PARAM_IN         = 'in';
    public const PARAM_NOTIN      = 'notin';
    public const PARAM_NULL       = 'null';
    public const PARAM_NOTNULL    = 'notnull';
    public const PARAM_NOT        = 'not';
    public const PARAM_REGEXP     = 'reg';

    /**
     * @var array<string,string|null>
     */
    public static $allowedOperators = [
        self::PARAM_EQ         => '=',
        self::PARAM_GT         => '>',
        self::PARAM_GTE        => '>=',
        self::PARAM_LT         => '<',
        self::PARAM_LTE        => '<=',
        self::PARAM_LIKE       => 'like',
        self::PARAM_LIKE_RIGHT => 'like',
        self::PARAM_LIKE_LEFT  => 'like',
        self::PARAM_IN         => null,
        self::PARAM_NOTIN      => null,
        self::PARAM_NULL       => null,
        self::PARAM_NOTNULL    => null,
        self::PARAM_NOT        => '!=',
        self::PARAM_REGEXP     => 'regexp',
    ];

    /**
     * @var \Mrcnpdlk\Lib\UrlSearchParser\Criteria\FilterParam[]
     */
    private $filters = [];

    /**
     * Filter constructor.
     *
     * @param array|FilterParam[] $filterArray
     *
     * @throws \Mrcnpdlk\Lib\UrlSearchParser\Exception\InvalidParamException
     *
     * @todo Check PLUS sign in string %2B code
     */
    public function __construct($filterArray = [])
    {
        if (!is_array($filterArray)) {
            throw new InvalidParamException(sprintf('FILTER params are invalid. Is [%s], Array expected', gettype($filterArray)));
        }

        /** @var FilterParam|string|array<string,string> $filters */
        foreach ($filterArray as $param => $filters) {
            if ($filters instanceof FilterParam) {
                $this->appendParam($filters);
                continue;
            }

            if (!is_string($param)) {
                throw new InvalidParamException(sprintf('Key in FILTER param is not a string'));
            }
            if (is_array($filters)) {
                /** @var string $operator */
                foreach ($filters as $operator => $value) {
                    if (!array_key_exists($operator, self::$allowedOperators)) {
                        throw new InvalidParamException(sprintf('Operator [%s] in FILTER is not allowed', $operator));
                    }
                    if (in_array($operator, [self::PARAM_IN, self::PARAM_NOTIN], true)) {
                        $value = '' === $value ? [] : explode(self::DELIMITER, $value);
                    }
                    if (in_array($operator, [self::PARAM_NULL, self::PARAM_NOTNULL], true)) {
                        $value = null;
                    }
                    $this->appendParam(new FilterParam($param, $operator, $value));
                }
            } elseif (is_string($filters)) {
                $this->appendParam(new FilterParam($param, self::PARAM_EQ, $filters));
            }
        }
    }

    /**
     * @param \Mrcnpdlk\Lib\UrlSearchParser\Criteria\FilterParam $filterParam
     *
     * @return \Mrcnpdlk\Lib\UrlSearchParser\Criteria\Filter<\Mrcnpdlk\Lib\UrlSearchParser\Criteria\FilterParam>
     */
    public function appendParam(FilterParam $filterParam): Filter
    {
        $this->filters[] = $filterParam;

        return $this;
    }

    /**
     * @param string $paramName
     *
     * @throws \Mrcnpdlk\Lib\UrlSearchParser\Exception\InvalidParamException
     *
     * @return \Mrcnpdlk\Lib\UrlSearchParser\Criteria\Filter<\Mrcnpdlk\Lib\UrlSearchParser\Criteria\FilterParam>
     */
    public function getByParam(string $paramName): Filter
    {
        $params = [];
        foreach ($this as $item) {
            if ($item->param === $paramName) {
                $params[] = $item;
            }
        }

        return new self($params);
    }

    /**
     * Retrieve an external iterator
     *
     * @see   http://php.net/manual/en/iteratoraggregate.getiterator.php
     *
     * @return Traversable<\Mrcnpdlk\Lib\UrlSearchParser\Criteria\FilterParam> An instance of an object implementing <b>Iterator</b> or
     *                                                                         <b>Traversable</b>
     *
     * @since 5.0.0
     */
    public function getIterator(): Traversable
    {
        return new ArrayIterator($this->filters);
    }

    /**
     * @param string $paramName
     *
     * @throws \Mrcnpdlk\Lib\UrlSearchParser\Exception\InvalidParamException
     *
     * @return \Mrcnpdlk\Lib\UrlSearchParser\Criteria\Filter Return removed part of filters
     */
    public function removeByParam(string $paramName): Filter
    {
        $i   = 0;
        $res = [];
        foreach ($this->filters as $item) {
            if ($item->param === $paramName) {
                $res[] = $item;
                unset($this->filters[$i]);
            }
            ++$i;
        }
        $this->filters = array_values($this->filters);

        return new self($res);
    }

    /**
     * @param \Mrcnpdlk\Lib\UrlSearchParser\Criteria\FilterParam $filterParam
     *
     * @return \Mrcnpdlk\Lib\UrlSearchParser\Criteria\Filter
     */
    public function replaceParam(FilterParam $filterParam): Filter
    {
        $isChanged = false;
        foreach ($this->filters as &$filter) {
            if ($filter->param === $filterParam->param && $filter->operator = $filterParam->operator) {
                $filter->value = $filterParam->value;
                $isChanged     = true;
            }
        }
        unset($filter);
        if (!$isChanged) {
            $this->appendParam($filterParam);
        }

        return $this;
    }

    /**
     * @return \Mrcnpdlk\Lib\UrlSearchParser\Criteria\FilterParam[]
     */
    public function toArray(): array
    {
        return $this->filters;
    }
}
